File: server/internal/expenses/bill_type.go
package expenses

import (
	"math/big"
	"time"

	"gorm.io/gorm"
)

type BillType struct {
	ID            uint           `json:"id" gorm:"primaryKey"`
	Name          string         `json:"name" gorm:"not null"`
	Icon          string         `json:"icon"`
	Color         string         `json:"color"`                                  // hex color code
	BillDay       int            `json:"bill_day"`                               // day of month, 0 means no specific day
	BillCycle     int            `json:"bill_cycle"`                             // in months, 0 means one-time
	FixedAmount   string         `json:"fixed_amount" gorm:"type:decimal(10,2)"` // Store as string for precision
	Stopped       bool           `json:"stopped" gorm:"default:false"`
	ExpenseTypeID *uint          `json:"expense_type_id" gorm:"index"` // optional, can be null
	UserID        uint           `json:"user_id" gorm:"not null;index"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `json:"-" gorm:"index"`
}

// GetFixedAmountAsBigFloat converts the stored string amount to big.Float
func (bt *BillType) GetFixedAmountAsBigFloat() *big.Float {
	amount := new(big.Float)
	amount.SetString(bt.FixedAmount)
	return amount
}

// SetFixedAmountFromBigFloat sets the amount from a big.Float
func (bt *BillType) SetFixedAmountFromBigFloat(amount *big.Float) {
	bt.FixedAmount = amount.String()
}

File: ios/backup/JiceotApp/Models/BillType.swift
//
//  BillType.swift
//  JiceotApp
//
//  Created on 12/11/2025.
//

import Foundation

struct BillType: Codable, Identifiable {
    let id: Int
    let name: String
    let icon: String
    let color: String
    let billDay: Int
    let billCycle: Int
    let fixedAmount: String?
    let stopped: Bool
    let expenseTypeId: Int?
    let userId: Int
    let createdAt: String
    let updatedAt: String
    
    enum CodingKeys: String, CodingKey {
        case id
        case name
        case icon
        case color
        case billDay = "bill_day"
        case billCycle = "bill_cycle"
        case fixedAmount = "fixed_amount"
        case stopped
        case expenseTypeId = "expense_type_id"
        case userId = "user_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct CreateBillTypeRequest: Codable {
    let name: String
    let icon: String?
    let color: String?
    let billDay: Int?
    let billCycle: Int?
    let fixedAmount: String?
    let expenseTypeId: Int?
    
    enum CodingKeys: String, CodingKey {
        case name
        case icon
        case color
        case billDay = "bill_day"
        case billCycle = "bill_cycle"
        case fixedAmount = "fixed_amount"
        case expenseTypeId = "expense_type_id"
    }
}

struct BillTypeListResponse: Codable {
    let billTypes: [BillType]
    let total: Int
    
    enum CodingKeys: String, CodingKey {
        case billTypes = "bill_types"
        case total
    }
}
